examples/10.dynamic_configuration/config_broker.cc

```diff
@@ -0,0 +1,275 @@
+
+	if (token == nullptr)
+	{
+		Debug::log("invalid config capability {}", sealed_cap);
+		return nullptr;
+	}
+
+	if (token->id == 0)
+	{
+		// Assign an ID so we can track the callbacks added
+		// from this capability
+		token->id = nextId++;
+
+		// Count how many config ids are defined to
+		// save us having to do this each time
+		for (auto i = 0; i < MAX_CONFIG_IDS; i++)
+		{
+			if (strlen(token->configId[i]))
+			{
+				token->count++;
+			}
+		}
+	}
+
+	return token;
+}
+
+//
+// Find a Config by name.  If it doesn't already exist
+// create one.
+//
+Config *find_config(const char *name)
+{
+	for (auto &c : configData)
+	{
+		if (strcmp(c->name, name) == 0)
+		{
+			return c;
+		}
+	}
+
+	// Allocate a Config object
+	Config *c = static_cast<Config *>(malloc(sizeof(Config)));
+
+	// Save the name
+	c->name = static_cast<char *>(malloc(strlen(name)));
+	strncpy(c->name, name, strlen(name));
+
+	c->updated = false;
+	c->data    = nullptr;
+
+	// Add it to the vector
+	configData.push_back(c);
+
+	return c;
+};
+
+//
+// Add a callback to the list for a Config.  Use the
+// id to prevent adding more than one callback for
+// each client
+//
+void add_callback(Config               *c,
+                  uint16_t              id,
+                  __cheri_callback void cb(const char *, void *))
+{
+	for (auto &cb_info : c->cbList)
+	{
+		if (cb_info->id == id)
+		{
+			cb_info->cb = cb;
+			return;
+		}
+	}
+
+	cbInfo *cb_info = static_cast<cbInfo *>(malloc(sizeof(cbInfo)));
+	cb_info->id     = id;
+	cb_info->cb     = cb;
+	c->cbList.push_back(cb_info);
+}
+
+//
+// Set a configuration data item.  The capabailty must have the
+// is_source set to true and include the name of the item being
+// set
+//
+void __cheri_compartment("config_broker")
+  set_config(SObj sealed_cap, const char *name, void *data)
+{
+	ConfigToken *token = config_capability_unseal(sealed_cap);
+
+	if (!token->is_source)
+	{
+		Debug::log("Not a source capability: {}", sealed_cap);
+		return;
+	}
+
+	bool valid = false;
+	for (auto i = 0; i < token->count; i++)
+	{
+		if (strcmp(token->configId[i], name) == 0)
+		{
+			valid = true;
+			break;
+		}
+	}
+	if (!valid)
+	{
+		Debug::log("Not a source capability for: {} {}", sealed_cap, name);
+		return;
+	}
+
+	// See if we already have a config structure
+	Config *c = find_config(name);
+
+	// Free any claim we had on the previous data
+	if (c->data)
+	{
+		heap_free(MALLOC_CAPABILITY, c->data);
+	}
+
+	// Create a read only version of the capability to pass to
+	// the subscribers, and add a claim so we can deliver it
+	// when a compartment registers a callback.
+	CHERI::Capability readOnlyData{data};
+	readOnlyData.permissions() &=
+	  {CHERI::Permission::Load, CHERI::Permission::Global};
+	c->data = readOnlyData;
+	heap_claim(MALLOC_CAPABILITY, c->data);
+
+	// Mark it as having been updated
+	c->updated = true;
+
+	pending++;
+	futex_wake(&pending, -1);
```

Comment:
Interesting;  Your right I hadn't thought about how the current approach trusts the sender not to change the data after the call to set_config().   I was trying to avoid the broker or sender having to copy the data , I was sort of thinking that in a real world example this might be some network receive buffer that could just be used to avoid the copy, but that could be impractical, and the config data sizes are likely to be small.

With this new approach is there any advantage it making the initial allocation against the provider's quota, if the broker is going to add a clam and the provider do a free (if fact II'm not sure when in this case the provider could call free ?) - or should the broker just allocate directly from it's quota and pass that as a non-global to the providers callback ?

Can you also please expand on what you meant by "Presently that set_config would again have to make-before-break the data allocation" - I didn't really follow that ?

Reply:
I think I would structure this code with a callback approach, as @nwf suggests, because then the broker is the only thing that needs to hold a mutable capability.

You can do a sequence of:

1. Broker allocates space
2. Broke invokes callback passing it the buffer as a local capability (remove Global permission), which guarantees no-capture for the mutable capability.
3. Publisher writes the data into the newly allocated buffer.
4. Publisher returns to broker.
5. Broker drops publisher's claim on the old object (but holds its own).
6. Broker attempts to claim new object with publisher's capability.
7. If this fails, broker claims the old version with the publisher's capability.  If this fails, the publisher is actively malicious, free the old value, publish null, return failure.  @nwf has a proposal for an allocator API that would make this atomic, which might be nice.
8. Broker updates the published pointer.
9. Broker frees the old pointer (dropping its own claim).
10. New pointer is now held by the broker, claimed by the broker and the publisher, and the broker has the *only* writeable capability to it.
11. Broker notifies all watchers, passing them a read-only capability.

If the watchers trust the broker, then they don't need to do checks on the pointers that they receive except to do `heap_claim` or `heap_claim_fast` if they want to check that it still exists.  They don't need to do either *during* their subscribe callback because the caller guarantees that the object remains live there.  

If they *don't* want to trust the broker, then they probably want to do `heap_claim_fast` and copy to prevent concurrent mutation.

It would be quite nice to showcase a normal watcher and a paranoid one.

In both cases, the watchers do not trust the publishers for anything related to memory safety (though they do trust that the thing that they get back is a valid value).

Reply:
> Can you also please expand on what you meant by "Presently that set_config would again have to make-before-break the data allocation" - I didn't really follow that ?

Sorry, I just mean that our allocator APIs would require you to perform `malloc()` (or `heap_claim()`) and check the result before calling `free()` on the old value.  This is unfortunate, because we'd really like the allocator to understand what we're doing and let us use the quota consumed by the old object to allocate the new one.  But that's a separate PR.

Reply:
> That of course bumps the number of trusted frames up, is that something we're generally fairly relaxed about ?

Yes, trusted stack frames are very small (24 bytes).

>  I'm not quire sure why the broker allocating a new space and have the publisher copy into that is a stronger model, but I feel I'm missing something ?

It depends a bit on the availability model.  If the allocation counts against the publisher's quota, then they can't use the broker's quota to exhaust memory.  If the broker does not claim it then the publisher can call `heap_free_all` to free it concurrently with other users.  This can also be used to allocate, then free from the publisher's side and trick the broker into exhausting memory. You can avoid the latter by allocating it with the token APIs, which require agreement from both the owner of the allocator quota and the owner of the sealing key to deallocate (these are omitted from `heap_free_all`).

> I'm assuming the publisher has no interest in the value once its passed it on (for example its just something that it received via the network)

Again, it depends on your availability model (and it's worth writing that up).  Without the publisher's claim there are two possibilities.  If the broker has a small quota, the total amount of published data is limited and individual publishers may be able to mount denial of service attacks on each other.  If the broker has a large quota, then individual publishers may publish huge things and exhaust total system memory.

You can potentially mitigate this by putting a maximum size for publishing in your capability that authorises publishing.  Then your auditing policy can specify both the total amount of published data and the amount of per-publisher data that can be stored.